### PR TITLE
kubelet: make the image pull time more accurate in event

### LINKED ISSUE #111704: When using --serialize-image-pulls=true "Successfully pulled image" Event outputs misleading time
### What happened?

While going over Kubernetes events I noticed that image pulls for some small images were taking several minutes according to the "Successfully pulled image" event. From investigating this is misleading as the actual pull time is much smaller but due to --serialize-image-pulls=true being the default in the observed cluster and other pods being initialized at the time, the reported pull time ends up accounting for time while waiting for other pods to finish pulling. 

### What did you expect to happen?

I expected to see the time it takes to download the image itself from a registry. Customers seem to rely on this metric for this information and we have received some complaints as a result where customers noted slow pull times as a result of misunderstanding the current behavior of this metric. 

### How can we reproduce it (as minimally and precisely as possible)?

Start with a K8s cluster with --serialize-image-pulls=true (default) - For my check I used default minikube v1.26.1

1. Create pod-large where pod-large pulls any large image (the image cannot be locally cached)
```
apiVersion: v1
kind: Pod
metadata:
  name: pod-large
spec:
  containers:
    - name: large-image
      image: roadmunk/largecontainer:latest
  restartPolicy: Never
```
2. Right after step 1 (within seconds) create pod-small where pod-small pulls any small image
```
apiVersion: v1
kind: Pod
metadata:
  name: pod-small
spec:
  containers:
    - name: small-image
      image: busybox:latest
  restartPolicy: Never
```

Check the output for the event after running: `kubectl describe pod pod-small`
```
other info ...
Events:
  Type    Reason     Age    From               Message
  ----    ------     ----   ----               -------
  Normal  Scheduled  7m41s  default-scheduler  Successfully assigned default/pod-small to minikube
  Normal  Pulling    7m39s  kubelet            Pulling image "busybox:latest"
  Normal  Pulled     2m40s  kubelet            Successfully pulled image "busybox:latest" in 4m59.26837101s
  Normal  Created    2m40s  kubelet            Created container small-image
  Normal  Started    2m39s  kubelet            Started container small-image
```

Note that despite the small size, pull time for the image in pod-small is several minutes, we should instead see it take seconds (which is how long the pull actually takes). To note this is the case you can try pulling another small image like alpine that isnt cached locally and compare after the above are done.

### Anything else we need to know?

Here is the code that pulls an image in kubernetes and logs the event:
https://github.com/kubernetes/kubernetes/blob/fbffe056dd03bd9c746e8819ad22043d640a4489/pkg/kubelet/images/image_manager.go#L156.

As can be seen from it, it's not true pull time, just time from invoking the puller to receiving a result.  The serial puller is using a Golang channel as a queue and will just wait until a previous image is done pulling as I noted before, thus it will end up recording the amount of time in the queue + pull time yielding a misleading metric. 

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:31:21Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:20:00Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Should reproduce anywhere (tried it on minikube) but have personally observed this in Azure as well
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
image pulling may hang for serial image pulling. The pull time is including waiting time for other images pulling. This is not accurate.

```
Successfully pulled image "docker.m.daocloud.io/library/ubuntu:22.10" in 49.18401355s
```
The time should be exactly how long it takes to pull this image. Or the event message should be something like 
```
Successfully pulled image "docker.m.daocloud.io/library/ubuntu:22.10" in 49.18401355s(72.871704s including waiting)
```

#### Which issue(s) this PR fixes:
Fixes #111704

#### Special notes for your reviewer:
- if a user needs to know the pending time, user can get the time between `Pulling` event and `Pulled` event.

**Before this change** the three pulling, the later one will include the time of the first pulling.
```
[root@paco ~]# kubectl create -f images/image.yaml
pod/pod-small-1 created
pod/pod-small-2 created
pod/pod-small-3 created
[root@paco ~]# kubectl get events -w| grep pulled
0s          Normal    Pulled                    pod/pod-small-3   Successfully pulled image "docker.m.daocloud.io/library/ubuntu:22.10" in 1m21.284220239s
0s          Normal    Pulled                    pod/pod-small-2   Successfully pulled image "docker.m.daocloud.io/library/ubuntu:20.04" in 2m7.876426944s
0s          Normal    Pulled                    pod/pod-small-1   Successfully pulled image "docker.m.daocloud.io/library/ubuntu:14.04" in 4m52.884660938s
```

**With this change** the later pulling will not take the first pulling time into account.
```
[root@paco ~]# kubectl create -f images/image.yaml
pod/pod-small-1 created
pod/pod-small-2 created
pod/pod-small-3 created
[root@paco ~]# kubectl get events -w| grep pulled
0s          Normal    Pulled             pod/pod-small-1                                 Successfully pulled image "docker.m.daocloud.io/library/ubuntu:14.04" in 1m10.319558171s
0s          Normal    Pulled             pod/pod-small-3                                 Successfully pulled image "docker.m.daocloud.io/library/ubuntu:22.10" in 27.63731662s
0s          Normal    Pulled             pod/pod-small-2                                 Successfully pulled image "docker.m.daocloud.io/library/ubuntu:20.04" in 31.496845543s
```

#### Does this PR introduce a user-facing change?
```release-note
None
```

### CODE REVIEW

**Comment by pacoxu** [comment]
@estebanreyl

**Comment by pacoxu** [comment]
the failure should be fixed by #111780

**Comment by yxxhero** [comment]
/assign 
The code is LGTM

**Comment by estebanreyl** [comment]
LGTM. Thanks again for looking at this 👍

File: pkg/kubelet/images/image_manager.go
Comment (endocrimes) [inline pkg/kubelet/images/image_manager.go:154]: Do we potentially want to include both? I'm a bit worried about losing the "whole chain" time, and the difference between the two could help surface other issues.

something like:

`fmt.Sprintf("Successfully pulled image %q in %v (%v including waiting)", container.Image, imagePullResult.pullDuration, time.Since(startTime))`

Comment (pacoxu) [reply pkg/kubelet/images/image_manager.go:154]: This would be more accurate. I will update the pr.

Comment (pacoxu) [reply pkg/kubelet/images/image_manager.go]: Updated in https://github.com/kubernetes/kubernetes/pull/111772/commits/dc7be1338e3667a3d77a4bd8942ba4e5187e0f8f.

**Review by endocrimes — APPROVED** [review]
/lgtm

**Comment by pacoxu** [comment]
Rebased.
@endocrimes would you lgtm again?

/assign @dchen1107 
for approval

**Comment by pacoxu** [comment]
Rebased.

This has been rebased for 4 times, 3times after LGTM 😓. Can some approvers take a look? 

/cc  @mrunalp  @klueska

**Comment by pacoxu** [comment]
rebased again😓

**Review by endocrimes — APPROVED** [review]
another
/lgtm

**Comment by bobbypage** [comment]
Can we consider to backport this to the recent few releases, e.g. 1.25, 1.24? I think it's a pretty small isolated change and will help to reduce a lot of confusion. 

I've seen quite a few folks being confused with the misleading events regarding image pulls, especially when used with serial image pulls. Often the case is that a small image pull may report a very long image pull time when in reality it was blocked behind a much larger image pull.

**Comment by njucjc** [comment]
> Can we consider to backport this to the recent few releases, e.g. 1.25, 1.24? I think it's a pretty small isolated change and will help to reduce a lot of confusion.
> 
> cc @pacoxu
> 
> I've seen quite a few folks being confused with the misleading events regarding image pulls, especially when used with serial image pulls. Often the case is that a small image pull may report a very long image pull time when in reality it was blocked behind a much larger image pull.

+1

**Comment by smarterclayton** [comment]
In observing events we have pretty high precision in the message:

 `Successfully pulled image "nginx:latest" in 4.614356447s (4.614439819s including waiting)`.

I was going to suggest we truncate the precision at a millisecond or 100 microseconds to shorten the message and make it a bit easier to read.  That would be `time.Since(startTime).Truncate(time.Millisecond)` instead of the exact length.

**Pull request merged**
